torchvision/prototype/transforms/functional/_geometry.py

```diff
@@ -1,72 +1,95 @@
-from typing import Tuple, List, Optional
+from typing import TypeVar, Any, cast
 
+import PIL.Image
 import torch
-from torchvision.transforms import (  # noqa: F401
-    functional as _F,
-    InterpolationMode,
-)
-
-horizontal_flip_image = _F.hflip
-
-
-def horizontal_flip_bounding_box(bounding_box: torch.Tensor, *, image_size: Tuple[int, int]) -> torch.Tensor:
-    shape = bounding_box.shape
-    bounding_box = bounding_box.view(-1, 4)
-    bounding_box[:, [0, 2]] = image_size[1] - bounding_box[:, [0, 2]]
-    return bounding_box.view(shape)
-
-
-_resize_image = _F.resize
-
+from torchvision.prototype import features
+from torchvision.prototype.transforms import kernels as K
+from torchvision.transforms import functional as _F
 
-def resize_image(
-    image: torch.Tensor,
-    size: List[int],
-    interpolation: InterpolationMode = InterpolationMode.BILINEAR,
-    max_size: Optional[int] = None,
-    antialias: Optional[bool] = None,
-) -> torch.Tensor:
-    new_height, new_width = size
-    num_channels, old_height, old_width = image.shape[-3:]
-    batch_shape = image.shape[:-3]
-    return _resize_image(
-        image.reshape((-1, num_channels, old_height, old_width)),
-        size=size,
-        interpolation=interpolation,
-        max_size=max_size,
-        antialias=antialias,
-    ).reshape(batch_shape + (num_channels, new_height, new_width))
+from ._utils import dispatch
 
+T = TypeVar("T", bound=features.Feature)
 
-def resize_segmentation_mask(
-    segmentation_mask: torch.Tensor,
-    size: List[int],
-    interpolation: InterpolationMode = InterpolationMode.NEAREST,
-    max_size: Optional[int] = None,
-    antialias: Optional[bool] = None,
-) -> torch.Tensor:
-    return resize_image(
-        segmentation_mask, size=size, interpolation=interpolation, max_size=max_size, antialias=antialias
-    )
 
+@dispatch(
+    {
+        torch.Tensor: _F.hflip,
+        PIL.Image.Image: _F.hflip,
+        features.Image: K.horizontal_flip_image,
+        features.BoundingBox: None,
+    },
+)
+def horizontal_flip(input: T, *args: Any, **kwargs: Any) -> T:
+    """ADDME"""
+    if isinstance(input, features.BoundingBox):
+        output = K.horizontal_flip_bounding_box(input, format=input.format, image_size=input.image_size)
+        return cast(T, features.BoundingBox.new_like(input, output))
+
+    raise RuntimeError
+
+
+@dispatch(
+    {
+        torch.Tensor: _F.resize,
+        PIL.Image.Image: _F.resize,
+        features.Image: K.resize_image,
+        features.SegmentationMask: K.resize_segmentation_mask,
+        features.BoundingBox: None,
+    }
+)
+def resize(input: T, *args: Any, **kwargs: Any) -> T:
+    """ADDME"""
+    if isinstance(input, features.BoundingBox):
+        size = kwargs.pop("size")
```

Comment:
Taking `resize_bounding_box` as an example, it would then take `size` and `old_image_size`. In that case we should probably also drop the "old" to make it `size` and `image_size`.